### ISSUE TITLE
Using JSURL in links

### ISSUE THREAD

**Issue by dkozar** [neherlab/covid19_scenarios#326]
# 🙋 Feature Request
Due to URL encoding, currently generated URLs are long and messy.

## 🔦 Context

We could use JSURL (https://github.com/Sage/jsurl) which would make the links shorter and even nicer.

With being shorter I mean: more data could be squeezed into this type of URL, without worrying of links being clipped by some systems.

## 😯 Describe the feature

Whenever the user changes parameters, we would update the route (browser link) and the parameters would be "JSURL encoded".

## 💻 Examples

JSON:

`{"name":"John Doe","age":42,"children":["Mary","Bill"]}`

JSON + URL encoding:

`%7B%22name%22%3A%22John%20Doe%22%2C%22age%22%3A42%2C%22children%22%3A%5B%22Mary%22%2C%22Bill%22%5D%7D`

JSURL:

`~(name~'John*20Doe~age~42~children~(~'Mary~'Bill))`

## 💁 Possible Solution

I used the JSURL before. The setup is trivial, here's the code snippet: https://github.com/dkozar/treelogy-demo/blob/09360f38c8974a57198eac7b65b0cdbaf5209c62/src/demo/utils/UrlUpdater.js

Here's the live example (some demo): https://dkozar.github.io/treelogy-demo/?data=~(versions~(~(id~%27Version_1~nodes~(~(id~%27foo~data~%27return*20*27foo*27*3b*20)~(id~%27bar~data~%27return*20*27bar*27*3b)))~(id~%27Version_2~parent~%27Version_1~nodes~(~(id~%27bar~data~%27return*20*27bar_2*27*3b)~(id~%27baz~data~%27return*20*27baz*27*3b)))))

Note that above link contains data for the demo.

![image](https://user-images.githubusercontent.com/1078403/78048873-4471dc00-7372-11ea-8464-e5d77b34240c.png)

Also note that, if backward compatibility with the old link is needed (for legacy purposes), our logic could check if the received link is of the old or the new format, and the app would be able to turn the old format into the new one.

**I could implement this feature.**

## Related

 - https://github.com/neherlab/covid19_scenarios/issues/31
 - https://github.com/neherlab/covid19_scenarios/issues/11
 - https://github.com/neherlab/covid19_scenarios/issues/284#issuecomment-606706469

**Comment by dkozar** [comment]
I've been playing with the implementation.

PR (work in progress): https://github.com/neherlab/covid19_scenarios/pull/329
My branch: https://github.com/dkozar/covid19_scenarios/tree/feat/jsurl

You can try it out locally. Once you Run the analysis, and refresh the page, the input parameters should be persisted. The system is backward compatible, meaning **it will load the old link format** too (URL encoded).

Note: I'm having some issues when reloading the web page and running again (my browser freezes). Will debug tomorrow.

**Comment by rcbevans** [comment]
Is there a specific reason you'd like to use jsurl?  Is a goal to keep the link human readable?  If the primary goal is shorten, a lzw/lzma base64 encoding of the data may be preferable, for example [json-url](https://www.npmjs.com/package/json-url)

**Comment by dkozar** [comment]
Yes, the primary goal is to shorten.

Btw the current link doesn't seem readable, too. Actually, it is less readable than JSURL: 
```
%7B%22name%22%3A%22John%20Doe%22%2C%22age%22%3A42%2C%22children%22%3A%5B%22Mary%22%2C%22Bill%22%5D%7D
```

Is there a goal to support the app responding to manual changes of the link? - I don't know. I personally wouldn't edit such a link by hand, not even sure about a motivation to do that.

Yes, we can use any encoder which compresses data into a string that's safe to transfer via query string.

But I like that JSURL gives at least a glympse of its data to the reader. Because, there will be a lot of copy-pasting of those links into the app (comparison mode between 2 scenarios?), or emails.

So, this:

```
~(name~'John*20Doe~age~42~children~(~'Mary~'Bill))
```

... to me seems much nicer than this:

```
woTCo29uZQHCo3R3bwLCpXRocmVlwpMBz 
```

**Comment by dkozar** [comment]
I fixed the browser freezing problem and wrote unit tests around the de/serialization:

![image](https://user-images.githubusercontent.com/1078403/78164903-6d61a200-7442-11ea-88d6-f7887e633a3e.png)

PR: https://github.com/neherlab/covid19_scenarios/pull/329

**Comment by dkozar** [comment]
PR: https://github.com/neherlab/covid19_scenarios/pull/329

Size comparison (URL Encoded vs JSURL):

![size](https://user-images.githubusercontent.com/1078403/78466067-09b2cf80-76f5-11ea-9942-255eeb97c573.png)

**Comment by ivan-aksamentov** [comment]
Resolved in #329
